crates/bevy_ecs/src/storage/table.rs

```diff
@@ -35,25 +35,52 @@ impl TableId {
 
     /// Creates a new [`TableId`].
     ///
-    /// `index` *must* be retrieved from calling [`TableId::index`] on a `TableId` you got
+    /// `index` *must* be retrieved from calling [`TableId::as_u32`] on a `TableId` you got
     /// from a table of a given [`World`] or the created ID may be invalid.
     ///
     /// [`World`]: crate::world::World
     #[inline]
-    pub fn new(index: usize) -> Self {
+    pub const fn from_u32(index: u32) -> Self {
+        TableId(index)
+    }
+
+    /// Creates a new [`TableId`].
+    ///
+    /// `index` *must* be retrieved from calling [`TableId::as_usize`] on a `TableId` you got
+    /// from a table of a given [`World`] or the created ID may be invalid.
+    ///
+    /// [`World`]: crate::world::World
+    ///
+    /// # Panics
+    ///
+    /// Will panic if the provided value does not fit within a [`u32`] with `debug_assertions`.
+    #[inline]
+    pub const fn from_usize(index: usize) -> Self {
+        debug_assert!((index as u128) < (u32::MAX as u128));
```

Comment:
`usize` is pointer sized, and we don't expect to have machines with 128 bit address space in decades to come.

But to avoid dealing with such questions I just prefer writing this code as:

```rust
debug_assert!(index as u32 as usize == index);
```

It will work correctly even in million years from now where pointer size is 256 bits.

Reply:
Just tested that this doesn't get optimized away, definitely like this check more, good idea!